scripts/timelapse-colorizer-data/generate_data.py

```diff
@@ -61,76 +64,51 @@
-    # we have 4 indices representing the bounds (2 sets of x,y coordinates).
-    # ushort can represent up to 65_535. Images with a larger resolution than this will need to replace the datatype.
-    bbox_data = np.zeros(shape=(totalIndices * 2 * 2), dtype=np.ushort)
 
     for group_name, frame in grouped_frames:
-        # take first row to get zstack path
-        row = frame.iloc[0]
-        frame_number = row["index_sequence"]
-
         start_time = time.time()
 
-        zstackpath = row["seg_full_zstack_path"]
+        # Get the path to the segmented zstack image frame from the first row (should be the same for
+        # all rows in this group, since they are all on the same frame).
+        row = frame.iloc[0]
+        frame_number = row[TIMES_COLUMN]
+        # Flatten the z-stack to a 2D image.
+        zstackpath = row[SEGMENTED_IMAGE_COLUMN]
         if platform.system() == "Windows":
             zstackpath = "/" + zstackpath
         zstack = AICSImage(zstackpath).get_image_data("ZYX", S=0, T=0, C=0)
         seg2d = zstack.max(axis=0)
-        mx = np.nanmax(seg2d)
-        mn = np.nanmin(seg2d[np.nonzero(seg2d)])
-        # float comparison with 1 here is okay because this is not a calculated value
-        if scale != 1.0:
-            seg2d = skimage.transform.rescale(
-                seg2d, scale, anti_aliasing=False, order=0
-            )
-        seg2d = seg2d.astype(np.uint32)
-
-        lut = np.zeros((mx + 1), dtype=np.uint32)
-        for row_index, row in frame.iterrows():
-            # build our remapping LUT:
-            label = int(row["label_img"])
-            rowind = int(row["initialIndex"])
-            lut[label] = rowind + lut_adjustment
 
-        # remap indices of this frame.
-        seg_remapped = lut[seg2d]
-
-        # Capture bounding boxes
-        # Optimize by skipping i = 0, since it's used as a null value in every frame
-        for i in range(1, lut.size):
-            # Boolean array that represents all pixels segmented with this index
-            cell = np.argwhere(seg_remapped == lut[i])
-
-            if cell.size > 0:
-                write_index = lut[i] * 4
-                # Reverse min and max so it is written in x, y order
-                bbox_min = cell.min(0).tolist()
-                bbox_max = cell.max(0).tolist()
-                bbox_min.reverse()
-                bbox_max.reverse()
-                bbox_data[write_index : write_index + 2] = bbox_min
-                bbox_data[write_index + 2 : write_index + 4] = bbox_max
+        # Scale the image and format as integers.
+        seg2d = writer.scale_image(seg2d)
+        seg2d = seg2d.astype(np.uint32)
 
-        # convert data to RGBA
-        seg_rgba = np.zeros(
-            (seg_remapped.shape[0], seg_remapped.shape[1], 4), dtype=np.uint8
+        # Remap the frame image so the IDs are unique across the whole dataset.
+        seg_remapped, lut = writer.remap_segmented_image(
```

Comment:
tiny naming nit: a writer isn't a thing that remaps images or scales them.

Reply:
Moved the remapping code and image scaling out of the writer class :)